java/ql/automodel/publish.sh

```diff
@@ -1,31 +1,118 @@
+  exit 1
+fi
+
+# Get the sha of the previous release
+PREVIOUS_RELEASE_SHA=$(git rev-list -n 1 main -- ./src/qlpack.yml)
+if [ -z "$PREVIOUS_RELEASE_SHA" ]; then
+  echo "Error: Could not get the sha of the previous release of codeml-automodel query pack"
+  exit 1
+else
+  echo "Previous query-pack release sha: $PREVIOUS_RELEASE_SHA"
+fi
+
+CURRENT_BRANCH=$(git rev-parse --abbrev-ref HEAD)
+CURRENT_SHA=$(git rev-parse HEAD)
+
+if [ -z "${1:-}" ]; then
+  # If the first argument is empty, use the latest release of codeml-automodel
+  TAG_NAME=$(gh api -H 'Accept: application/vnd.github+json' -H 'X-GitHub-Api-Version: 2022-11-28' /repos/github/codeml-automodel/releases/latest | jq -r .tag_name)
+  # Check TAG_NAME is not empty
+  if [ -z "$TAG_NAME" ]; then
+      echo "Error: Could not get latest release of codeml-automodel"
+      exit 1
+  fi
+  echo "Updating to latest automodel release: $TAG_NAME"
+  rm release.zip || true
+  gh release download $TAG_NAME -A zip -O release.zip --repo 'https://github.com/github/codeml-automodel'
+  rm -rf release || true
+  unzip -o release.zip -d release 
+  REVISION=$(jq -r '.["codeql-sha"]' release/codeml-automodel*/codeml-automodel-release.json)
+  echo "The latest codeml-automodel release specifies the codeql sha $REVISION"
+  # Check that REVISION is downstream from PREVIOUS_RELEASE_SHA
+  if ! git merge-base --is-ancestor "$PREVIOUS_RELEASE_SHA" "$REVISION"; then
+    echo "Error: The codeql version $REVISION is not downstream of the query-pack version $PREVIOUS_RELEASE_SHA"
+    exit 1
+  fi
+  # Get the version of the codeql code specified by the codeml-automodel release
+  git checkout "$REVISION"
+else
+  # Check that the current HEAD is downstream from PREVIOUS_RELEASE_SHA
+  if ! git merge-base --is-ancestor "$PREVIOUS_RELEASE_SHA" "$CURRENT_SHA"; then
+    echo "Error: The current HEAD is not downstream from the previous release"
+    exit 1
+  fi
+fi
 
 AUTOMODEL_ROOT="$(readlink -f "$(dirname $0)")"
 WORKSPACE_ROOT="$AUTOMODEL_ROOT/../../.."
 GRPS="automodel,-test"
 
-if [ -z "$CODEQL_DIST" ]; then
-  echo "CODEQL_DIST not set"
-  exit -1
-fi
-
-cd "$AUTOMODEL_ROOT"
+pushd "$AUTOMODEL_ROOT"
 echo Testing automodel queries
 "${CODEQL_DIST}/codeql" test run test
+popd
 
-cd "$WORKSPACE_ROOT"
+pushd "$WORKSPACE_ROOT"
+echo "Preparing the release"
+"${CODEQL_DIST}/codeql" pack release --groups $GRPS -v
 
-echo Preparing release
-"${CODEQL_DIST}/codeql" pack release --groups $GRPS
+echo "Publishing the release"
+# Add --dry-run to test publishing
+"${CODEQL_DIST}/codeql" pack publish --groups $GRPS -v
+
+echo "Bumping versions"
+"${CODEQL_DIST}/codeql" pack post-release --groups $GRPS -v
+popd
+
+# The above commands update
+#  ./src/CHANGELOG.md
+#  ./src/codeql-pack.release.yml
+#  ./src/qlpack.yml
+# and add a new file
+#  ./src/change-notes/released/<version>.md
+
+if [ -z "${1:-}" ]; then
+  # If we used the latest release of codeml-automodel, then we need to return to the current branch
```

Comment:
Could we expand that comment to remind us it's fine to do so because we worked from the last commit that modified those files - and so our changes won't be conflicting with checking out the branch

Reply:
Here we're simply returning to our working branch (after publishing a different version of the repo specified by a release `sha`). The script has already ensured that you do not have any locally modified and uncommitted files. So returning to the working branch will work fine -- but you will now have some modified files that need to be committed.